WebDriverAgentTests/IntegrationTests/FBTouchActionIntegrationTests.m

```diff
@@ -0,0 +1,388 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import <XCTest/XCTest.h>
+
+#import "FBIntegrationTestCase.h"
+
+#import "XCUIElement.h"
+#import "XCUIApplication+FBTouchAction.h"
+#import "FBAlert.h"
+#import "FBTestMacros.h"
+#import "XCUIDevice+FBRotation.h"
+#import "FBRunLoopSpinner.h"
+
+@interface FBTouchActionIntegrationTestsPart1 : FBIntegrationTestCase
+@end
+
+@interface FBTouchActionIntegrationTestsPart2 : FBIntegrationTestCase
+@property (nonatomic) XCUIElement *pickerWheel;
+@end
+
+
+@implementation FBTouchActionIntegrationTestsPart1
+
+- (void)verifyGesture:(NSArray<NSDictionary<NSString *, id> *> *)gesture orientation:(UIDeviceOrientation)orientation
+{
+  [[XCUIDevice sharedDevice] fb_setDeviceInterfaceOrientation:orientation];
+  NSError *error;
+  XCTAssertTrue(self.testedApplication.alerts.count == 0);
+  [self.testedApplication fb_performTouchAction:gesture error:&error];
+  FBAssertWaitTillBecomesTrue(self.testedApplication.alerts.count > 0);
+}
+
+- (void)setUp
+{
+  [super setUp];
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    [self launchApplication];
+    [self goToAlertsPage];
+  });
+}
+
+- (void)tearDown
+{
+  [super tearDown];
+  [[FBAlert alertWithApplication:self.testedApplication] dismissWithError:nil];
+}
+
+- (void)testErroneousGestures
+{
+  NSArray<NSArray<NSDictionary<NSString *, id> *> *> *invalidGestures =
+  @[
+    // Empty chain
+    @[],
+    
+    // Chain element without 'action' key
+    @[@{
+        @"options": @{
+            @"ms": @100
+            }
+        },
+      ],
+    
+    // Empty chain because of cancel
+    @[@{
+        @"action": @"tap",
+        @"options": @{
+            @"element": self.testedApplication.buttons[FBShowAlertButtonName],
+            }
+        },
+      @{
+        @"action": @"wait",
+        @"options": @{
+            @"ms": @100
+            }
+        },
+      @{
+        @"action": @"cancel"
+        },
+      ],
+    
+    // Chain with unknown action
+    @[@{
+        @"action": @"tapP",
```

Comment:
Does this mean `tap`?

Reply:
this is intended to fail - the test verifies expected errors in case of invalid chains